qiskit/transpiler/passes/optimization/hoare_opt.py

```diff
@@ -0,0 +1,340 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+""" Pass for hoare logic circuit optimization. """
+
+import sys
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.circuit import ControlledGate, Gate
+from qiskit.extensions.unitary import UnitaryGate
+from qiskit.quantum_info.operators.predicates import matrix_equal
+from . import _gate_extension  # pylint: disable=W0611
+try:
+    from z3 import And, Or, Not, Implies, Solver, Bool, unsat
+except ModuleNotFoundError:
+    print("""Z3 package not found, which is required to run this pass.
+             Please install Z3 via 'pip install z3-solver'.""")
+    sys.exit(1)
```

Comment:
This should be inside a try except block and handled gracefully for the places where z3 is used. See: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/passes/optimization/crosstalk_adaptive_schedule.py#L36-L40 for an example. This way we're only requiring people install z3 when someone is using the `HoareOptimizer` pass. Right now the way this is written python will exit on import when z3 which is not the correct behavior.

Reply:
Thank you for bringing this up. Will fix.